### PR TITLE
User send request

### PR DESCRIPTION
**Pull request by maiq721**

### CODE REVIEW

**Comment by maiq721** [comment]
ready

**Comment by maiq721** [comment]
https://dev.framgia.com/people/2654

File: Gemfile [hunk 1]
```diff
@@ -24,6 +24,7 @@ gem "sass-rails", "~> 5.0"
 gem "sqlite3"
 gem "turbolinks", "~> 5"
 gem "uglifier", ">= 1.3.0"
+gem "pry"
```
Comment (mdmahmudurrahman) [inline Gemfile]: Please add this gem in the development group

File: app/assets/javascripts/carts.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    if (quantity <= 0 ) $("#update_item" + id + " > .cart_quantity > input").val(1);
+  $.ajax({
+      url:"/carts/update_quantity",
+      type:'post',
+      async: true,
+      beforeSend: function(xhr) {
+          xhr.setRequestHeader('X-CSRF-Token', $('meta[name="csrf-token"]').attr('content'))
+      },
+      data:$('#update_item'+id).serialize(),
+  })
```
Comment (mdmahmudurrahman) [inline app/assets/javascripts/carts.js]: Please use single quote in JS and check the indentation

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -14,9 +16,25 @@ def load_user
...
+
   def logged_in_user
     return if logged_in?
     flash[:danger] = t ".not_login"
     redirect_to login_url
   end
+
+  def load_cart
+    @cart = current_cart
+    flash[:danger] = t(".cart_not_found") if @cart.nil?
```
Comment (mdmahmudurrahman) [inline app/controllers/application_controller.rb]: Please remove optional ()

File: app/controllers/carts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  def update_quantity
+    session["cart"][params[:key]] = params[:quantity].to_i
+    @book_in_cart = Book.by_book_ids session["cart"].keys
+    respond_to do |format|
+      format.js
+    end
+  end
+
+  def destroy
+    session["cart"] = nil if session["cart"]
```
Comment (mdmahmudurrahman) [inline app/controllers/carts_controller.rb]:
```ruby 
session.delete :cart if session[:cart]
```

File: app/controllers/carts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    redirect_to root_path
+  end
+
+  def remove_book
+    session["cart"].delete(@book.id.to_s) if session["cart"].key? @book.id.to_s
```
Comment (mdmahmudurrahman) [inline app/controllers/carts_controller.rb]: Remove ()

File: app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  def index
+    @orders = Order.all
+  end
+
+  def new
+    if session["cart"].empty?
+      flash[:danger] = t ".empty"
+      redirect_to root_path
+    end
```
Comment (mdmahmudurrahman) [inline app/controllers/orders_controller.rb]: Add this block to a filter for this action

File: app/controllers/orders_controller.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    @order = Order.new order_params
+    @order.total = total_money
+    @flag = ActiveRecord::Base.transaction do
+      @order.save!
+      session["cart"].each do |key, val|
+        @order_detail = OrderDetail.create!(book_id: key.to_i,
+          order_id: @order.id, quantity: val, current_price: price_book(key.to_i))
+      end
+    end
+    session[:cart] = nil
```
Comment (mdmahmudurrahman) [inline app/controllers/orders_controller.rb]: `session.delete :cart `

File: app/controllers/orders_controller.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    end
+    @order = Order.new
+  end
+
+  def create
+    @order = Order.new order_params
+    @order.total = total_money
+    @flag = ActiveRecord::Base.transaction do
+      @order.save!
+      session["cart"].each do |key, val|
```
Comment (mdmahmudurrahman) [inline app/controllers/orders_controller.rb]: `session[:cart]`

File: app/controllers/orders_controller.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        @order_detail = OrderDetail.create!(book_id: key.to_i,
+          order_id: @order.id, quantity: val, current_price: price_book(key.to_i))
```
Comment (mdmahmudurrahman) [inline app/controllers/orders_controller.rb]: Remove redundant ()

File: app/helpers/carts_helper.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    session["cart"][book.id.to_s]
+  end
+
+  def load_book_in_cart
+  	if session["cart"].blank?
+  	  flash[:danger] = t ".danger"
+      redirect_to books_path
+  	else
+  	  @book_in_cart = Book.by_book_ids session["cart"].keys
+  	end
```
Comment (mdmahmudurrahman) [inline app/helpers/carts_helper.rb]: Indentation

File: app/helpers/orders_helper.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+module OrdersHelper
+  def price_book key
+    Book.find_by(id: key).price
+  end
+
+  def name_item book_id
+    Book.find_by(id: book_id).name
+  end
+
+
```
Comment (mdmahmudurrahman) [inline app/helpers/orders_helper.rb]: Remove empty line

File: app/models/book.rb [hunk 12]
```diff
@@ -9,6 +9,8 @@ class Book < ApplicationRecord
 
   scope :list_book, ->{select :id, :name, :image, :price,
                         :category_id, :created_at}
+  scope :by_book_ids, ->(book_ids){where(id: book_ids)}
```
Comment (mdmahmudurrahman) [inline app/models/book.rb]: Please remove ()

File: app/models/book.rb [hunk 13]
```diff
@@ -9,6 +9,8 @@ class Book < ApplicationRecord
...
+  scope :by_id, ->(book_id){where(id: book_id)}
```
Comment (mdmahmudurrahman) [inline app/models/book.rb]: The same here

File: app/views/carts/_table.html.erb [hunk 14]
```diff
@@ -0,0 +1,34 @@
+<table class="table">
+  <tr>
+    <th>name</th>
+    <th>price</th>
+    <th>quantity</th>
+    <th>subtotal</th>
+    <th>action</th>
```
Comment (mdmahmudurrahman) [inline app/views/carts/_table.html.erb]: Indentation

File: app/views/carts/_table.html.erb [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  <% @book_in_cart.each do |book| %>
+    <tr>
+      <td><%= book.name %></td>
+      <td><%= book.price %></td>
+      <td>
+        <div class="cart-quantity">
+          <%= form_tag update_quantity_carts_path,method: :post, remote: true, id: "update_item"+book.id.to_s  do |f| %>
+            <input type="text" name="key" hidden value="<%= book.id %>">
+            <div class="form-group">
+            <%= number_field_tag :quantity, quantity(book), class: "form-control", id: "a#{book.id.to_s}", onchange: "changeQuantity(#{book.id.to_s})", within: 1...10 %>
```
Comment (mdmahmudurrahman) [inline app/views/carts/_table.html.erb]: Maximum 80 characters per line

File: app/views/carts/_table.html.erb [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+            </div>
+          <% end %>
+        </div>
+      </td>
+      <td><%= sub_total(book) %></td>
+      <td>
+        <div class="btn btn-default">
+          <%= button_to remove_book_path(book), remote: true, method: :delete,  data: { confirm: t(".confirm_del") }, class: "delete" do  %>
+            Remove
```
Comment (mdmahmudurrahman) [inline app/views/carts/_table.html.erb]: I18n

File: app/views/carts/_table.html.erb [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+            <div class="form-group">
+            <%= number_field_tag :quantity, quantity(book), class: "form-control", id: "a#{book.id.to_s}", onchange: "changeQuantity(#{book.id.to_s})", within: 1...10 %>
+            </div>
+          <% end %>
+        </div>
+      </td>
+      <td><%= sub_total(book) %></td>
+      <td>
+        <div class="btn btn-default">
+          <%= button_to remove_book_path(book), remote: true, method: :delete,  data: { confirm: t(".confirm_del") }, class: "delete" do  %>
```
Comment (mdmahmudurrahman) [inline app/views/carts/_table.html.erb]: 1. Maximum 80 characters per line
2. Remove while space after { and before }

File: app/views/orders/index.html.erb [hunk 18]
```diff
@@ -0,0 +1,37 @@
+ <% provide :title, t(".title") %>
+ <h2><%= t ".your_order" %></h2>
+
+ <div class="row full-screen">
+   <% if @orders %>
+   <div class="order-summary">
```
Comment (mdmahmudurrahman) [inline app/views/orders/index.html.erb]: Add 2 space indentation

File: app/views/orders/index.html.erb [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+              <td><%= detail.quantity %></td>
+              <td><%= total_each_order_detail(detail) %></td>
+            </tr>
+          </tbody>
+          <% end %>
+        </table>  
+        <div class="order-col">
+          <div><strong><%= t ".total_money" %> <%= order.total %> </strong></div>
+        </div>
+	    <% end %>
```
Comment (mdmahmudurrahman) [inline app/views/orders/index.html.erb]: Indentation

File: app/views/orders/index.html.erb [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    </div>
+  </div>
+  <% else %>
+	<p><%= t ".p" %></p>
```
Comment (mdmahmudurrahman) [inline app/views/orders/index.html.erb]: Indentation

File: app/views/orders/new.html.erb [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+        <%= render "shared/error_messages", object: f.object %>
+        <div class="detail">
+	      <div class="form-group">
+	        <%= f.text_field :full_name, placeholder: "Full name", class: "input form-control" %>
+	      </div>
+          <div class="form-group">
+            <%= f.text_field :phone, placeholder: "Phone number", class: "input form-control" %>
+          </div>
+          <div class="form-group">
+            <%= f.text_field :address, placeholder: "Address", class: "input form-control" %>
```
Comment (mdmahmudurrahman) [inline app/views/orders/new.html.erb]: For the place-holders, add i18n

File: app/helpers/orders_helper.rb [hunk 22]
```diff
@@ -0,0 +1,14 @@
+module OrdersHelper
+  def price_book key
+    Book.find_by(id: key).price
+  end
+
+  def name_item book_id
+    Book.find_by(id: book_id).name
+  end
```
Comment (mdmahmudurrahman) [inline app/helpers/orders_helper.rb]: 1. You can see both of the methods are doing the same. So try to write a dynamic method to perform this query. 
2. Write this method on the model.

**Comment by mdmahmudurrahman** [comment]
commented

File: app/controllers/carts_controller.rb [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+  def add_cart
+    key_book = @book.id.to_s
+    return redirect_to carts_path if session["cart"].key? key_book
+    session["cart"][key_book] = Settings.cart.quantity
+    redirect_to carts_path
+  end
+
+  def update_quantity
+    session["cart"][params[:key]] = params[:quantity].to_i
```
Comment (mdmahmudurrahman) [inline app/controllers/carts_controller.rb]: `session[:cart]`

File: app/controllers/carts_controller.rb [hunk 24]
```diff
@@ -0,0 +1,31 @@
+class CartsController < ApplicationController
+  before_action :load_cart, :load_book, only: %i(add_cart remove_book update_quantity)
+  before_action :load_book_in_cart, only: :index
+
+  def index; end
+
+  def add_cart
+    key_book = @book.id.to_s
+    return redirect_to carts_path if session["cart"].key? key_book
```
Comment (mdmahmudurrahman) [inline app/controllers/carts_controller.rb]: `session[:cart]`

File: app/controllers/carts_controller.rb [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+    session["cart"][key_book] = Settings.cart.quantity
```
Comment (mdmahmudurrahman) [inline app/controllers/carts_controller.rb]: The same here

File: app/controllers/carts_controller.rb [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+    redirect_to carts_path
+  end
+
+  def update_quantity
+    session["cart"][params[:key]] = params[:quantity].to_i
+    @book_in_cart = Book.by_book_ids session["cart"].keys
```
Comment (mdmahmudurrahman) [inline app/controllers/carts_controller.rb]: The same here

**Comment by maiq721** [comment]
ready

**Pull request merged**